clients/pkg/promtail/targets/docker/target_group.go

```diff
@@ -0,0 +1,139 @@
+package docker
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/docker/docker/client"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/positions"
+	"github.com/grafana/loki/clients/pkg/promtail/targets/target"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const DockerSource = "Docker"
+
+// targetGroup manages all container targets of one Docker daemon.
+type targetGroup struct {
+	metrics       *Metrics
+	logger        log.Logger
+	positions     positions.Positions
+	entryHandler  api.EntryHandler
+	defaultLabels model.LabelSet
+	host          string
+	client        client.APIClient
+
+	mtx     sync.Mutex
+	targets map[string]*Target
+}
+
+func (tg *targetGroup) sync(groups []*targetgroup.Group) {
+	tg.mtx.Lock()
+	defer tg.mtx.Unlock()
+
+	for _, group := range groups {
+		if group.Source != DockerSource {
+			continue
+		}
+
+		for _, t := range group.Targets {
+			containerID, ok := t[dockerLabelContainerID]
```

Comment:
You should be applying relabeling here allowing users to rename/drop/keep discovered labels